### PR TITLE
Introduce Amazon Block Storage  Manager (EBS)

### PR DESCRIPTION
**Pull request by gberginc**
This commit sets up the core structure of the new AWS EC2 block storage
manager. It provides just the basic stubs and tests that are to be
extended to allow for complete management of Amazon block storage

The refresher collects data about existing volumes and private
snapshots. Volumes are not yet linked to disks of virtual machines
collected by the cloud manager.

Depends on [Add storage manager mixins](https://github.com/ManageIQ/manageiq/pull/13384) PR from core repo.

@miq-bot add_label enhancement

### CODE REVIEW

**Comment by gberginc** [comment]
@aiperon please review

**Comment by gberginc** [comment]
@agrare could https://github.com/ManageIQ/manageiq/commit/d557f190a406dc7671fe246716806f93db1933b5 be affecting travis tests?

```
     Failure/Error:
       def initialize(inventory)
         @inventory             = inventory
         @inventory_collections = inventory.inventory_collections
       end
     
     ArgumentError:
       wrong number of arguments (given 2, expected 1)
```

Is there any doc on how to approach the refresh parsers now?

**Comment by agrare** [comment]
@gberginc yeah it is failing for one of my branches using `InventoryObjectRefresh` refresh also.
@Ladas can you change `RefreshParserInventoryObject` so it doesn't have to use the experimental inventory abstraction?

**Comment by gberginc** [comment]
Now that you mentioned another branch, I checked also other PRs in Amazon repo and noticed the one from @Ladas #102.

I guess it's best to wait until #102 is passing and then ensure storage manager is using the same approach.

**Comment by Ladas** [comment]
@gberginc the CI should be fixed, but I've caused a conflict in the specs, you'll need to rebase

**Comment by gberginc** [comment]
@Ladas thanks. The conflict is simple to resolve as just the number of `ext_management_system`s is different. However I guess it would be better to use the same approach for storage manager as it is for cloud/network. I'll try to update the PR.

Another thing I wanted to ask: I currently only use stubbed specs, but there are also those that are using actual AWS account - how should I approach this as using our account will change the VCR as well as all the existing spec data.

**Comment by Ladas** [comment]
@gberginc cool

So the VCR kinda sucks there, because we have a VCR env, that we use for that. So option would be that I would record it for you, once you are ready, or we would share some readonly creds for the VCR env. So for now, I think easiest will be that I will record and send the VCRs, once you have all the API calls in place.

We were thinking about getting away from the VCRs and just use stubbed specs, but then it's nice to have a spec that is using real env. data.

**Comment by gberginc** [comment]
@Ladas I've just added additional commit for the inventory object support. I am not sure if you are the one to review this PR (if not, please let me know so that I can ask someone else).

Regarding VCR recording: do I understand correctly, I should now add `EmsRefresh.refresh(@ems.block_storage_manager)` everywhere VCR cassettes are used, for example:

https://github.com/ManageIQ/manageiq-providers-amazon/blob/master/spec/models/manageiq/providers/amazon/cloud_manager/refresher_inventory_object_spec.rb#L38

Then you run the spec and share the VCR with me so that I can update the actual tests?

**Comment by Ladas** [comment]
@gberginc I'll do the review

Regarding the VCR, yes, lets do the VCR specs as a next PR

**Comment by sergio-ocon** [comment]
Hey @Ladas, did you review it?

**Comment by gberginc** [comment]
I think this is partly dependent on the current changes to the refresh parsers, https://github.com/ManageIQ/manageiq-providers-amazon/pull/105.

I would nevertheless appreciate feedback to continue working on storage managers.

File: app/models/manageiq/providers/amazon/block_storage_manager/refresh_parser_inventory_object.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    $aws_log.info("#{log_header}...}")
+    get_volumes
+    get_snapshots
+    $aws_log.info("#{log_header}...Complete")
+
+    inventory_collections
+  end
+
+  private
+  def get_volumes
```
Comment (aiperon) [inline app/models/manageiq/providers/amazon/block_storage_manager/refresh_parser_inventory_object.rb]: We're using the same parsing code for RefreshParser and RefreshParserInventoryObject.
Wouldn't it be better if we move it into ManageIQ::Providers::Amazon::RefreshHelperMethods ?

Comment (Ladas) [inline app/models/manageiq/providers/amazon/block_storage_manager/refresh_parser_inventory_object.rb]: there is a slight difference in the parsing code. But yeah, we can abstract the parsing part. But that can be done in another PR.

But my goal is to deprecate/delete the old parsing code and use the new one, so maybe it won't be needed. :-)

File: app/models/manageiq/providers/amazon/block_storage_manager/refresh_parser_inventory_object.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    def volume_type
+      "ManageIQ::Providers::Amazon::BlockStorageManager::CloudVolume"
+    end
+
+    def volume_snapshot_type
+      "ManageIQ::Providers::Amazon::BlockStorageManager::CloudVolumeSnapshot"
+    end
+  end
+
+
```
Comment (Ladas) [inline app/models/manageiq/providers/amazon/block_storage_manager/refresh_parser_inventory_object.rb]: extra blank lines

File: app/models/manageiq/providers/amazon/block_storage_manager/refresh_parser_inventory_object.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    $aws_log.info("#{log_header}...}")
+    get_volumes
+    get_snapshots
+    $aws_log.info("#{log_header}...Complete")
+
+    inventory_collections
+  end
+
+  private
```
Comment (Ladas) [inline app/models/manageiq/providers/amazon/block_storage_manager/refresh_parser_inventory_object.rb]: blank line before and after private

File: app/models/manageiq/providers/amazon/cloud_manager.rb [hunk 4]
```diff
@@ -54,6 +72,14 @@ def ensure_managers
     network_manager.provider_region = provider_region
   end
 
+  def ensure_block_storage_managers
+    build_block_storage_manager unless block_storage_manager
+    block_storage_manager.name            = "#{name} Block Storage Manager"
+    block_storage_manager.zone_id         = zone_id
+    block_storage_manager.provider_region = provider_region
+    true
```
Comment (Ladas) [inline app/models/manageiq/providers/amazon/cloud_manager.rb]: why we return true?

File: app/models/manageiq/providers/amazon/cloud_manager.rb [hunk 5]
```diff
@@ -54,6 +72,14 @@ def ensure_managers
     network_manager.provider_region = provider_region
   end
 
+  def ensure_block_storage_managers
+    build_block_storage_manager unless block_storage_manager
```
Comment (Ladas) [inline app/models/manageiq/providers/amazon/cloud_manager.rb]: I think we can just put this body to ensure_managers. Or calling this from the body of ensure_managers, then we should extract the current into :ensure_network_manager, then add :ensure_block_storage_manager

**Comment by Ladas** [comment]
@gberginc just few nits, mostly rubocop that should be fixed. After that I am ready to merge, so we can move forward. :-)

File: app/models/manageiq/providers/amazon/cloud_manager.rb (hunk 5 above)
Comment (gberginc) [inline app/models/manageiq/providers/amazon/cloud_manager.rb]: @Ladas I've merged the two into `ensure_managers` as there is currently no need to have them separated. I also fixed the other rubocop issues.

**Comment by gberginc** [comment]
@Ladas I amended the commit, so the following comment is not visible. Just reposting to make sure it doesn't get lost.

I've merged the two into ensure_managers as there is currently no need to have them separated. I also fixed the other rubocop issues - I think the remaining issues can be left as they are following existing code.

**Comment by Ladas** [comment]
@gberginc awesome :-) Thank you for a great contribution. 

Yeah, the rest of the robocop warnings are on the edge of a correctness. :-)

**Comment by sergio-ocon** [comment]
👍

**Comment by roliveri** [comment]
@gberginc @Ladas @Fryguy  Actually, this shouldn't have been merged yet. We've been naming managers based on concrete names instead of type. This enables providers to support different/multiple implementations of the same type of manager. So, I would suggest a name change.


From:
```ruby
ManageIQ::Providers::Amazon::BlockStorageManager < ManageIQ::Providers::StorageManager
```
To:
```Ruby
ManageIQ::Providers::Amazon:: StorageManager::EBS < ManageIQ::Providers::StorageManager
```

**Comment by gberginc** [comment]
@roliveri do you think we could proceed with the pending PRs on the manageiq repo and then introduce this change now that it's merged?

**Comment by roliveri** [comment]
@gberginc Is there a benefit in doing that? It seems that 2 of the pending PRs rely on the same naming scheme:

https://github.com/ManageIQ/manageiq/pull/13458
https://github.com/ManageIQ/manageiq/pull/13457

So I would make the same comment on those. If I merge those, you'll only have to submit new PRs updating those as well.

**Comment by gberginc** [comment]
@roliveri the only consideration was that `ManageIQ/manageiq` was failing tests because of this merge. I though it would be faster to make the existing PRs merged and then do a followup patch that would fix the names.

**Comment by roliveri** [comment]
@gberginc Ah, i see. It looks like https://github.com/ManageIQ/manageiq/pull/13458 has been merged, and doing so has fixed the test failures.

So, https://github.com/ManageIQ/manageiq/pull/13457 remains. I can hold off merging that until the names are fixed, or I can merge it, but then I'd like for you to open a Github issue for the name change - just to make sure it doesn't get overlooked.

let me know which way you'd like to go.

**Comment by gberginc** [comment]
@roliveri I already started on renaming the class. I therefore suggest you wait with the worker as it is not required for functional test.

**Pull request merged**